code/__HELPERS/jatum.dm

```diff
@@ -0,0 +1,221 @@
+// JATUM. System for making the serializing and deserializing of PoD types easier
+
+/**
+  * Gets the flat list that can be passed in a `new /type(argslist(retval))` expression to recreate the datum. Must only return a list containing values that can be JATUM serialized
+  */
+/datum/proc/jatum_new_arglist()
+	return null
+
+/// The JATUM version used for serializing and deserializing
+#define JATUM_VERSION 1
+
+/**
+  * Attempt to serialize a given value to the JATUM format.
+  *
+  * * value - The var to serialize.
+  */
+/world/proc/jatum_serialize(value)
+	var/list/json_structure
+	try
+		json_structure = _jatum_serialize_value(value, list())
+	catch(var/exception/e)
+		CRASH(e)
+
+	return json_encode(list(
+		"jatum\\version" = JATUM_VERSION,
+		"content" = json_structure
+	))
+
+/world/proc/_jatum_serialize_value(value, list/seen_references)
+	if(isnull(value))
+		return null
+
+	if(isnum(value) || istext(value))
+		return list(
+			"type" = "jatum\\raw",
+			"value" = value
+		)
+
+	if(ispath(value))
+		return list(
+			"type" = "jatum\\path",
+			"path" = value
+		)
+
+	var/ref = "\ref[value]"
+	var/existing_ref = seen_references[ref]
+	if(existing_ref)
+		return list(
+			"type" = "jatum\\ref",
+			"jatum\\id" = existing_ref)
+
+	// Simple incrementing ID system
+	var/ref_id = seen_references.len + 1
+	seen_references[ref] = ref_id
+
+	if(istype(value, /matrix)) // matricies work with json_encode so use a custom arglist
+		return list(
+			"jatum\\id" = ref_id,
+			"type" = "/matrix",
+			"jatum\\new_arglist" = _jatum_serialize_value(json_decode(json_encode(value))), seen_references)
+
+	if(istype(value, /savefile)) // Just store the path, rely on BYOND for the rest
+		var/savefile/save = value
+		return list(
+			"jatum\\id" = ref_id,
+			"type" = "/savefile",
+			"jatum\\new_arglist" = _jatum_serialize_value(list(save.name), seen_references)
+		)
+
+	if(istype(value, /regex)) // store construct params
+		var/regex/rx = value
+		return list(
+			"jatum\\id" = ref_id,
+			"type" = "/regex",
+			"name" = rx.name,
+			"flags" = rx.flags
+		)
+
+	if(islist(value))
+		// Serialize all lists as dicts, list("a") and list("a" = null) can't be differentiated in DM
+		var/list_contents = list()
+		for(var/key in value)
+			var/got_l_value
+			var/l_value
+			try
+				l_value = value[key]
+				got_l_value = TRUE
+			catch
+				// Expected, indicates a flat list
```

Comment:
```suggestion
			var/got_l_value = FALSE
			var/l_value = null
			if(!isnum(key))
				try
					l_value = value[key]
					got_l_value = TRUE
				catch
					// Expected, indicates a flat list
```
You can have list-out-of-bounds runtimes if you try to use `value[key]` on a list such as `list(21, 43, 4)`

Reply:
That's what the catch guard is for

Reply:
ah, alright